File: src/008/main.c
# include <stdbool.h>
# include <stdlib.h>
# include <stdio.h>

typedef unsigned long u64;
typedef unsigned char u08;

int main (int argc, char * args []) {
	const int N_DIGITS = 13;
	u08 * current = malloc (N_DIGITS * sizeof (u08));
	u64 largest = 0;
	
	for (int i = 0; i < N_DIGITS; i++) {
		current [i] = 0;
	}
	
	FILE * file = fopen (args [1], "r");
	for (int index = 0; true; index = (index + 1) % N_DIGITS) {
		char c = fgetc (file);
		
		if (c == '\n' || c == EOF) {
			break;
		}
		
		current [index] = c - 48;
		
		u64 product = 1;
		for (u64 i = 0; i < N_DIGITS; i++) {	
			product *= current [i];
		}
		
		if (product > largest) {
			largest = product;
		}
	}
	
	fclose (file);
	free (current);
	
	printf ("%lu\n", largest);
}

File: src/008/main.rs
use std::io::BufReader;
use std::io::BufRead;
use std::fs::File;

fn main () {
	const N_DIGITS: usize = 13;
	let mut current: [u8; N_DIGITS] = [0; N_DIGITS];
	let mut largest: u64 = 0;
	
	let args: Vec<String> = std::env::args ().collect ();
	let buffer = BufReader::new (File::open (&args [1]).unwrap ());
	
	let mut index: usize = 0;
	for line in buffer.lines () {
		for c in line.unwrap ().chars () {
			current [index] = c.to_digit (10).unwrap () as u8;
			index = (index + 1) % N_DIGITS;
			
			let mut product: u64 = 1;
			for i in current {
				product *= i as u64;
			}
			
			if product > largest {
				largest = product;
			}
			
			index = (index + 1) % N_DIGITS;
		}
	}
	
	println! ("{}", largest);
}

